src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -390,48 +406,113 @@ internal unsafe int Read(byte[] array, int offset, int count, int timeout)
+            _readQueue.Enqueue(result);
 
-                if ( (events & Interop.Sys.PollEvents.POLLIN) == 0)
-                {
-                    throw new TimeoutException();
-                }
+            return result.UnderlyingTask;
+        }
+
+        public override Task WriteAsync(byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (_inBreak)
+                throw new InvalidOperationException(SR.In_Break_State);
+            if (array == null)
+                throw new ArgumentNullException(nameof(array));
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedPosNum);
+            if (count < 0)
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedPosNum);
+            if (array.Length - offset < count)
+                throw new ArgumentException(SR.Argument_InvalidOffLen);
+            if (count == 0)
+                return Task.CompletedTask; // return immediately if no bytes to write; no need for overhead.
+
+            // check for open handle, though the port is always supposed to be open
+            if (_handle == null)
+                throw new ObjectDisposedException(SR.Port_not_open);
+
+            Memory<byte> buffer = new Memory<byte>(array, offset, count);
+            SerialStreamAsyncResult result = new SerialStreamAsyncResult(cancellationToken, buffer);
+            _writeQueue.Enqueue(result);
+            return result.UnderlyingTask;
+        }
+
+        public override IAsyncResult BeginRead(byte[] array, int offset, int numBytes, AsyncCallback userCallback, object stateObject)
+        {
+            if (array == null)
+                throw new ArgumentNullException(nameof(array));
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedNonNegNumRequired);
+            if (numBytes < 0)
+                throw new ArgumentOutOfRangeException(nameof(numBytes), SR.ArgumentOutOfRange_NeedNonNegNumRequired);
+            if (array.Length - offset < numBytes)
+                throw new ArgumentException(SR.Argument_InvalidOffLen);
+
+            if (_handle == null)
+                throw new ObjectDisposedException(SR.Port_not_open);
+
+            Memory<byte> buffer = new Memory<byte>(array, offset, numBytes);
+            return BeginReadCore(SerialPort.InfiniteTimeout, buffer, userCallback, stateObject);
+        }
+
+        // Will wait `timeout` miliseconds or until reading or writing is possible
+        // If no operation is requested it will wait.
+        // Returns event which has happened
+        private Interop.Sys.PollEvents PollEvents(int timeout, bool pollReadEvents, bool pollWriteEvents)
+        {
+            if (!pollReadEvents && !pollWriteEvents)
+            {
+                Thread.Sleep(timeout);
```

Comment:
How long is this timeout going to be generally?  It'd be great if we only kept the worker alive while it was needed.

Reply:
This specific timeout (for poll) is always 1ms, for reading operations it is infinity but it is fairly common scenario to change it (for reading at least). I'll explore the idea of doing this but I think I will address it separately because I do not want the loop to start 6 time if you do 6 synchronous Read calls in a row (perhaps add how much time at most the loop can stay idle before it gets killed - let's solve it separately as it will be easier to see the changes and focus on this specific problem)

Reply:
Ok.  Let's make sure to follow-up, though.  This is going to burn a thread that sits polling every millisecond for as long as the stream is alive, even if it's just sitting idle.

Reply:
ok this is done as part of this PR since we got dispose issue with this